### PR TITLE
[WIP] Move gproc related code into an 'adapter'

### PR DESCRIPTION
**Pull request by joshk**
The end goal is to allow for a Redis adapter for event propagation (instead gproc).
This will allow users to have the flexibility of using Redis when running multi tenant setups.

Yet to add:
- [ ] method and module docs for `Gproc`
- [ ] method and module docs for `Registry`
- [ ] test the `GProc` module
- [x] remove `:gproc` from all the test files
- [x] either lock to Elixir 1.2 and above, or have Travis test more combos

### CODE REVIEW

File: lib/poxa/adapters/gproc.ex [hunk 1]
```diff
@@ -0,0 +1,65 @@
+defmodule Poxa.Adapter.GProc do
+  @behavour Poxa.Registry
```
Comment (edgurgel) [inline lib/poxa/adapters/gproc.ex]: I think it's ` @behaviour`

File: lib/poxa/registry.ex [hunk 2]
```diff
@@ -0,0 +1,92 @@
+defmodule Poxa.Registry do
+
```
Comment (edgurgel) [inline lib/poxa/registry.ex]: I would probably just define the private function `adapter` and directly use

`adapter.register(channel)`

File: lib/poxa/adapters/gproc.ex [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  end
+
+  def user_count(channel) do
+    find_users(channel)
+    |> Enum.count
+  end
+
+  def occupied?(channel) do
+    match = {{:p, :l, {:pusher, channel}}, :_, :_}
+    select_count([{match, [], [true]}]) != 0
```
Comment (edgurgel) [inline lib/poxa/adapters/gproc.ex]: This can probably reuse `subscription_count(channel) != 0`

**Comment by joshk** [comment]
thanks for the feedback @edgurgel, incorporated and pushed

**Comment by joshk** [comment]
I have three instances of `gproc` to move into the `GProc` adapter, sooooooo close!

**Comment by joshk** [comment]
@edgurgel all `gproc` code as been moved into `GProc` with the full test suite passing

could you please review the changes so far?

next up is to improve the test suite to not use `gproc` everywhere, and to test the `GProc` adapter itself.

any and all advice welcomed :)

File: lib/poxa/registry.ex [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  def subscriptions do
+    adapter.subscriptions
+  end
+
+  def clean_up do
+    adapter.clean_up
+  end
+
+  @callback register(String.t) :: any
```
Comment (edgurgel) [inline lib/poxa/registry.ex]: I would probably use bang! for these calls like `register!`, `unregister!` `send_events!`

Comment (joshk) [inline lib/poxa/registry.ex]: cool, I'll get these changed

**Comment by arrowcircle** [comment]
@joshk why you need redis adapter?
Erlang/Elixir have build in cluster functionality. Gproc also support cluster (global registry) mode.

**Comment by joshk** [comment]
@arrowcircle setting up gproc to work across hosts is not super simple, while redis provides a simpler config for easier scaling.

We require Poxa to run across hosts, and since Redis is already a major part of our infrastructure it seems like a reasonable opt-in.

**Comment by edgurgel** [comment]
@arrowcircle I'm fine supporting Redis as an adapter as well as pushing forward the distributed gproc set up. 
The Registry will be a nice extraction either way as we had `gproc` calls every where and it was hard to maintain.

I would also like to run benchmarks to see how the performance is affected having a runtime configuration for the adapter. If it's too costly we may need to enforce it during compile time so the release will be tied to adapter selected at that time.

**Comment by arrowcircle** [comment]
@edgurgel @joshk Sure distributed erlang is much complicated compared to redis and few single nodes.
FYI https://github.com/doubleyou/dps distributed pub sub erlang project with full replication.

**Comment by edgurgel** [comment]
@arrowcircle I'm pretty confident that having just the nodes with gproc will be a lot faster than having to communicate with Redis. I still bet that distributed gproc will be the best solution if we find a nice way to coordinate them and keep them connected. 

If somebody is interested I would love to see how Rabbitmq and Riak deal with this and how people use Docker as well with them.

**Comment by edgurgel** [comment]
And BTW even if/when we have a Redis adapter I would use Redis just for what needs to be "distributed" across nodes but still using gproc to publish locally. So Redis would simply "aggregate" counters, sets of `user_ids` and do what needs to be done to relay messages. Publishing locally should still simply use gproc.

**Comment by joshk** [comment]
That is a very nice idea! @edgurgel would you have some time this week to meet for a coffee and talk about what is next to get this stuff merged in (hints and pointers) and then maybe we can talk about the redis adapter a bit?

**Comment by edgurgel** [comment]
Sure! Let's keep in touch.

**Comment by joshk** [comment]
Ok, so I have added a round of module docs, and some syntax bits, maybe all this work is not wanted, or over the top.

I also now test against other Erlang and Elixir versions, but the odd thing is the flapping tests. Any idea what I'm doing wrong?

Next up is adding some tests for the GProc Registry, and some docs for the Registry.

**Comment by edgurgel** [comment]
Ignore the flappy tests! I will try to fix on master.

**Comment by edgurgel** [comment]
Also please do not change stuff that is not related to the PR. Do you mind removing credo and everything that can't be considered part of this work. I would like to keep everything as small as possible and not prolong too much. We already got 32 files changed :/

 Cheers!

**Comment by Hengjie** [comment]
This is really cool stuff! I'd love to learn about what the next steps are after this PR to get it to something we can use on production (or maybe this PR is stable enough to use on production today?). I'm currently using Slanger in distributed setup with Redis, and we're on the same boat as @joshk; with an existing Redis infrastructure.

**Comment by iurifq** [comment]
I really like this idea, any progress on this @joshk? If you cannot finish it now, I'd be glad to take on this task of moving gproc code into an adapter.

**Comment by iurifq** [comment]
We implemented this idea in #74. Now we can plug new adapters to poxa :muscle:

**Pull request closed without merging**